File: SaranyaScalaActors/src/MobyDivideConquer.scala

/**
 * Name : Saranya Radhakrishnan
 * CWID : 20062589
 * Problem : Parallelize reading of Moby Dick files and get count of top 10 unique words. 
 * Compare the results for uni-processor and multiprocessors.
 * Also Compare the results with java code
 *
 */

/** Import java packages */
import java.lang;
import java.io._;

/** Import scala packages */
import scala.collection.mutable._;
import scala.actors.Actor._;
import scala.io._

/**
 * This object demonstrates Moby divide-conquer process in Scala.
 */
object MobyDivideConquer {

      /**
       * Entry point to the program 
       * Contains all the code required for Parent Thread
       */
    	def main(args : Array[String]) : Unit ={
       try{
    			    val TIMEOUT = 50000
    
              //Start time for the first loop
              val lintStartTime : Double = System.currentTimeMillis()
              
    					//Step 1 : Get the number of processors
    					val lintNoOfCores = Runtime.getRuntime.availableProcessors
    
    					//Step 2 : Read Directory of files into a list
    					val lstrFileNames : List[String] = getFiles("../moby-38205/")//("./Data/");
    
              //Check if the directory is not incorrect
              if(lstrFileNames == null)
              {
                println("Incorrect directory. Please enter the correct directory")
                exit()
              } 
              
    			    //Total number of files for processing
    			    val lintTotalLength = lstrFileNames.length
    					// Used to receive message from the child actors
    					val caller = self
    					//Used to increment the range
    					val lintDifference = (lintTotalLength.toDouble/lintNoOfCores).toDouble.ceil.toInt
    
              
    					//Step 3 : Parallelize the code for number of cores
    					(0 to lintNoOfCores - 1).foreach { i => 
    					      // Lower limit of the file to be processed by the child actor 
    					      val lower = i * lintDifference 
    
    					      // Upper limit of the file to be processed by the child actor
    					      val upper = (lintTotalLength - 1) min (( i + 1 ) * lintDifference - 1)
    					      
                    //Spawn the child actor
    					      actor {
    						        caller ! processFiles(lstrFileNames,lower, upper)
    					      }
    					} 
    
    					//Step 4: Process the output
    					processOutput(lintNoOfCores, "Core", lintStartTime, lintNoOfCores, lintTotalLength)
              
              if(lintTotalLength < 200 )
              {
                      //Start time for the second loop
            					val lintSecondStartTime : Double = System.currentTimeMillis()
            
            					//Step 5 : Parallelize the code for number of files
            					(0 until lintTotalLength).foreach { i  =>
            					        // Lower limit of the file to be processed by the child actor 
            					        val lower = i  
            					        
                              //Spawn the child actor
            					        actor {
            						         caller ! processFiles(lstrFileNames,lower, lower)
            					        }
            					} 
            
            					//Step 6 : Process the output
            					processOutput(lintTotalLength, "File", lintSecondStartTime, lintNoOfCores, lintTotalLength) 
              }
       }
       catch{
           case e : Exception => println(e.getMessage)
       }
    	}

      
      /**
       * Receives the result from each child actor
       * And merges all the partial Hash maps into a final hash map.
       * It also displays all the output parameters.
       * 
       * @param pintLimitForLoop
       *            : Contains the number of expected results for which the parent is to wait.
       *            
       * @param pstrTypeOfLoop
       *            : Contains the lower index from which the files are to be processed.
       *  
       * @param pintStartTime
       *            : Contains the time at which the entire file processing started.
       *  
       *  @return pintNumOfCores
       *            : Contains the number of cores utilized for processing the files.
       */
    	def processOutput(pintLimitForLoop : Int, pstrTypeOfLoop : String, pintStartTime : Double , pintNumOfCores : Int, pintTotalLength : Int) : Unit ={
    			try{
        				//Step 4 and 6 : Receive partial Hashmaps from child actors
        				val lhashParentResult  = (0 until pintLimitForLoop ).foldLeft(HashMap[String, Int]()) {
                      (lhashPartialResult, lintIndex) => {
                    					receive {
                    						    //Receive message of type Hashmap
                          					case lobjhashReceived : HashMap[_, _] => {
                            						val lhashReceived = lobjhashReceived.asInstanceOf[HashMap[String, Int]]
                            								for(lkeyValuePair <- lhashReceived){
                            									    val count = lhashPartialResult.getOrElse(lkeyValuePair._1,0)
                            											lhashPartialResult(lkeyValuePair._1) = count + lhashReceived(lkeyValuePair._1)
                            								}
                            						lhashPartialResult
                          					}
                                    case null => {
                                         println("Child "+ lintIndex +"'s Hashmap received is null")
                                         null
                                    }
                    				  }
            				 }
        				}
    
        				// End time for both type of processes : File and Core
        				val lintEndTime  : Double = System.currentTimeMillis()
    						// Total Wait time for both type of processes : File and Core
    						val lintWaitTime : Double = (lintEndTime - pintStartTime)/ 1000.0; 
    
        				//Check if the parent hashmap is not null
                if(lhashParentResult == null)
                {
                    println("Result Hashmap is null.")
                    exit()
                }
                
                //Step 7 : Sort the final hashmap
    				    val llstSortedHash = lhashParentResult.toSeq.sortBy(_._2).reverse.toList
    
    						println("\n============ OUTPUT (for "+ pstrTypeOfLoop +") ============= ") 
    
    						//Step 8: Display top 10 results
    						for(count <- 0  until  10)
    							println(llstSortedHash(count)._1 + "\t"+ llstSortedHash(count)._2)
    
    
  							//Step 9 : Display the parameters like Efficiency , Speedup , etc..
  							println("\nNumber of cores, N : "+ pintNumOfCores);
  						  println("Wait Time, dt :  "+ lintWaitTime);
  
  							val lintTimeForT1 = pintNumOfCores match {
                                          								case 1 => {
                                          									lintWaitTime
                                          								} 
    
                                          								case 4 =>   {
                                          									pstrTypeOfLoop match {
                                                    									case "Core" => {
                                                                                         if( pintTotalLength == 13635)
                                                                                           16.143
                                                                                         else if(  pintTotalLength == 38205 )
                                                                                           155.907
                                                                                         else
                                                                                           1.952
                                                                                      }  
                                                    									case "File" => 0.398
                                          									}
                                          								} 
                                                          
                                          								case 8 =>   {
                                          									pstrTypeOfLoop match {
                                          									case "Core" => 0.372
                                          									case "File" => 0.334
                                          									}
                                          								} 
    								                                      
                                                          case _ =>  
                                                            println("Unexpected number of cores!!")
                                                            1
                }
    
								println("Wait Time for 1 core, T1 :" + lintTimeForT1)

								if( lintWaitTime == 0 ) 
								{
  									val lintR = lintTimeForT1 / 1
  									println("R = t1/t"+pintNumOfCores + " = "+ lintR);
  									println("Efficiency, e: "+ lintR/pintNumOfCores);
								}
								else
								{
  									val lintR = lintTimeForT1 / lintWaitTime
  									println("Speedup, R :"+ lintR);
  									println("Efficiency, e = R/N  = " + lintR/pintNumOfCores);
								}
    			}
          catch{
    			    case e : Exception => println(e.getMessage)
    			}
    	}


      /**
       * Process each file in the range of Lower to Upper in the List of Files 
       * It creates a dictionary or Hashmap of all unique words and their count 
       * as found in the files.
       * It also returns this hashmap back to the calling function.
       * 
       * @param plstFileName
       *            : Contains the list of file paths.
       *            
       * @param pintLower
       *            : Contains the lower index from which the files are to be processed.
       *  
       * @param pintUpper
       *            : Contains the upper index upto which the files are to be processed.
       *  
       *  @return lhashChildResult
       *            : Contains the hashmap of all unique words and their count 
       *              from all the files processed by the actor.
       */
    	def processFiles(plstFileName : List[String], pintLower: Int, pintUpper: Int): HashMap[String, Int] ={
        try{
        			//Declare a local Hashmap for the actor
        			val lhashChildResult : HashMap[String, Int] = new HashMap[String, Int]()
        
        					//For every file in the range Lower -> Upper
        					(pintLower to pintUpper).foreach{ lintFileIndex => {
        						// Step1 : Read the file content line by line
        						val llstLines = Source.fromFile(plstFileName(lintFileIndex)).getLines.toList
        
        								//Step2: for every line in the Lines array -> split it into words
        								llstLines.foreach{ lstrLine =>
        								val llstWords = lstrLine.trim.toLowerCase.split(" +").toList
        
        								// Step3: For every word, check if it exists in the hashmap
        								//if exists, increase the counter
        								// else insert into the hashmap
        								llstWords.foreach{ lstrWord => {
        									if(lstrWord.trim != "")
        									{
        										val count = lhashChildResult.getOrElse(lstrWord.trim.toLowerCase, 0)
        												lhashChildResult(lstrWord.trim.toLowerCase) = count + 1
        									}
        								}
        								}
        						}
        					}
        	}
        	lhashChildResult
        }
        catch{
             case e : Exception => println(e.getMessage) 
             null
        }
    	}
    
      /**
       * Fetches the file paths for all files from a given directory 
       * And returns it to the calling function
       * 
       * @param pstrDirectory
       *            : Contains the parent directory path.
       *  
       *  @return 
       *            : Contains a List of file paths
       */
    	def getFiles(pstrDirectory: String): List[String] =  {
        try{
              //Fetch the files from the given directory.
			        val lfileRootDir = new File(pstrDirectory) 
              //Save the file paths in a list and return it to the calling function
			        lfileRootDir.listFiles.filter(_.isFile).map { lfile=> lfile.getPath }.toList
          }
          catch{
               case e : Exception => println(e.getMessage) 
               null
          }
	    }
}

File: SaranyaParallelThreads/src/MobyDivideConquer.java

/*
 * Name : Saranya Radhakrishnan
 * CWID : 20062589
 * Problem : Parallelize reading of Moby Dick files and get count of every unique words. 
 * Compare the results for uni-processor and multiprocessors
 *
 */

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

/**
 * This class demonstrates Moby divide-conquer process.
 */
public class MobyDivideConquer implements Runnable {
	public static final int TIMEOUT = 50 * 1000; // used to ensure thread is not
													// waiting for infinite
													// amount of time
	private static int lintIdCounter = 0; // Global id counter for child threads

	// Class variable used for queue of file paths
	private ArrayList<String> larrLstFilePaths = null;
	// Class variable for child dictionary
	private HashMap<String, Double> larrLstWordCount = new HashMap<String, Double>();
	// Class variable used for result queue
	private ArrayList<HashMap<String, Double>> larrLstResult = new ArrayList<HashMap<String, Double>>();

	private String lstrFilePath = ""; // Path of file for this thread
	private int lintChildId = -1; // Contains id for each child thread

	/**
	 * Entry point to the program Contains all the code required for Parent
	 * Thread
	 **/
	public static void main(String[] args) {
		try {
			// Declaring all the local variable

			// Local variable used for queue of file paths
			ArrayList<String> larrLstFilePaths = new ArrayList<>();
			// Local variable used for result queue
			ArrayList<HashMap<String, Double>> larrLstResults = new ArrayList<HashMap<String, Double>>();
			// Local variable for Final parent dictionary
			HashMap<String, Double> larrLstWordCount = new HashMap<String, Double>();
			// Temporary variable for dequeuing from result queue
			HashMap<String, Double> larrLstTempCount = new HashMap<String, Double>();

			// Step 1: Fetch the file paths from the directory in an ArrayList
			final File lFolders = new File("../15-text");

			if (!lFolders.isDirectory()) {
				System.out.println("Path for files does not exist");
				System.exit(0);
			}
			for (File lFile : lFolders.listFiles()) {
				String lstrPath = lFile.getPath();
				larrLstFilePaths.add(lstrPath);
			}

			System.out.println("parent waiting for child...");

			// Set the start time for the entire process
			long llngStartTime = System.currentTimeMillis();
			// Number of cores in the system
			int lintNoOfCores = 1;//Runtime.getRuntime().availableProcessors();

			// Step 2: Launch the child worker threads equivalent to the cores
			Thread child[] = new Thread[lintNoOfCores];
			for (int lintChildId = 0; lintChildId < lintNoOfCores; lintChildId++) {
				MobyDivideConquer lMDCThread = new MobyDivideConquer(larrLstFilePaths, larrLstResults);
				(child[lintChildId] = new Thread(lMDCThread)).start();

			}

			// Step 3: Wait for the child threads to finish processing and push
			// the results
			int lintChildDone = 0;
			while (lintChildDone < lintNoOfCores) {
				{
					synchronized (larrLstResults) {
						try {
							// Wait but not indefinitely
							larrLstResults.wait(TIMEOUT);

							int size = larrLstResults.size();
							// Step 4: Fetch the result HashMap from the queue
							while (larrLstResults.size() > 0) {
								larrLstTempCount = larrLstResults.remove(0);
								for (String key : larrLstTempCount.keySet()) {
									if (larrLstWordCount.containsKey(key))
										larrLstWordCount.put(key,
												larrLstWordCount.get(key) + larrLstTempCount.get(key));
									else
										larrLstWordCount.put(key, larrLstTempCount.get(key));
								}
							}
							lintChildDone += size;
						} catch (Exception ex) {
							System.out.println(ex.getMessage());
						}
					}
				}
			}

			// Contains the end time for processing
			long llngEndTime = System.currentTimeMillis();
			// Total duration of the process
			double ldblWaitTime = (llngEndTime - llngStartTime) / 1000.0;

			// Step 5: Sort the contents of the Hashmap
			ArrayList<WordCount> larrLstSorted = new ArrayList<WordCount>();

			// Convert the Hashmap into a list
			for (String lstrWord : larrLstWordCount.keySet())
				larrLstSorted.add(new WordCount(lstrWord, larrLstWordCount.get(lstrWord)));

			// Sort the list based on the count
			Collections.sort(larrLstSorted);
			int count = 0;
			System.out.println("\nCores, N = " + lintNoOfCores);
			System.out.println("Time Waiting, dT = " + ldblWaitTime);
			double t1 = 0;
			switch (lintNoOfCores) {
			case 1:
				t1 = 0.234;
				break;
			case 2:
				t1 = 0.2;
				break;
			case 4:
				t1 = 0.203;
				break;
			case 12:
				t1 = 0.219;
				break;
			}

			System.out.println("t1 = " + t1);
			System.out.println("R = t1/t"+lintNoOfCores + " = " + t1 / ldblWaitTime);
			System.out.println("e = R/N  = " + (t1 / ldblWaitTime) / lintNoOfCores);

			System.out.println("\n******  TOP 10 WORDS  *****\n");
			for (WordCount lWordCount : larrLstSorted) {
				if (count < 10) {
					System.out.println(lWordCount.lstrWord + "  \t  " + lWordCount.ldblCount);
					count++;
				} else
					break;
			}
		} catch (Exception ex) {
			System.out.println(ex.getMessage());
		}
	}

	/**
	 * Constructor Maps the parent variables to the child variables to create
	 * shared queues
	 * 
	 * @param parrLstFilePaths
	 *            : Contains the path to every file in the folder
	 * @param parrLstHashResult
	 *            : Contains the hashmap of every child thread
	 */
	public MobyDivideConquer(ArrayList<String> pArrLstFilePaths, ArrayList<HashMap<String, Double>> pArrLstHashResult) {
		try {
			this.lintChildId = lintIdCounter++;
			this.larrLstFilePaths = pArrLstFilePaths;
			this.larrLstResult = pArrLstHashResult;
		} catch (Exception e) {
			System.out.println(e.getMessage());
		}
	}

	/**
	 * Transfers here when thread started
	 */
	@Override
	public void run() {
		try {
			// Step 1: Loop until all the files have been processed
			while (true) {
				lstrFilePath = "";

				// / ENTERING critical region to safely manipulate the work
				// queue
				synchronized (larrLstFilePaths) {
					// Step 2 : If queue is not empty
					// Dequeue a filename from the queue
					// If no files remaining, run a sleeping thread.
					int lintSize = larrLstFilePaths.size();
					if (lintSize != 0)
						lstrFilePath = larrLstFilePaths.remove(0);
				}
				// / EXITING critical region

				// If no work, we're done, otherwise do work OUTSIDE the
				// critical region
				// as we have our work and there no point risking deadlock.
				// Step 3: Check if File path is not Empty
				if (lstrFilePath.isEmpty())
					break;

				// Step 4 : Process the file for which the path is received
				doWork();
			}

			// Step 5: Push the individual hashmap into the queue and notify the
			// parent thread
			synchronized (larrLstResult) {
				try {
					larrLstResult.add(larrLstWordCount);
					larrLstResult.notify();
				} catch (Exception ex) {
					System.out.println(ex.getMessage());
				}
			}
			System.out.println("Child " + lintChildId + " is done!");
		} catch (Exception ex) {
			System.out.println(ex.getMessage());
		}
	}

	/**
	 * Process each file to create hashmaps
	 * 
	 */
	private Integer doWork() {
		try {
			FileInputStream lfileInputStreamIn = new FileInputStream(lstrFilePath);
			BufferedReader lbuffReader = new BufferedReader(new InputStreamReader(lfileInputStreamIn));
			String lstrLine = "";

			while ((lstrLine = lbuffReader.readLine()) != null) {
				// Step 2: Split each line into words and insert into an array
				String[] lstrArrWords = lstrLine.toLowerCase().split(" +");

				// Step 3: While inserting into the dictionary, check if the key
				// already exists
				// If exist, update the count
				// Else add the word and count as 1 into the dictionary
				for (int i = 0; i < lstrArrWords.length; i++) {
					if (!lstrArrWords[i].trim().isEmpty()) {
						if (larrLstWordCount.containsKey(lstrArrWords[i]))
							larrLstWordCount.put(lstrArrWords[i], larrLstWordCount.get(lstrArrWords[i]) + 1);
						else
							larrLstWordCount.put(lstrArrWords[i], 1.0);
					}
				}
			}
			lbuffReader.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (Exception e) {
			System.out.println(e.getMessage());
		}
		return 0;
	}
}

class WordCount implements Comparable<WordCount> {
	public final String lstrWord;
	public final Double ldblCount;

	public WordCount(String pstrWord, Double pdblCount) {
		this.lstrWord = pstrWord;
		this.ldblCount = pdblCount;
	}

	@Override
	public int compareTo(WordCount p) {
		return (int) (p.ldblCount - this.ldblCount);
	}
}